File: samples/java/src/main/java/com/alibaba/cloud/cms/samples/client/AgentClient.java
package com.alibaba.cloud.cms.samples.client;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

import com.aliyun.auth.credentials.Credential;
import com.aliyun.auth.credentials.provider.StaticCredentialProvider;
import com.aliyun.sdk.gateway.pop.Configuration;
import com.aliyun.sdk.gateway.pop.auth.SignatureVersion;
import com.aliyun.sdk.service.starops20260428.AsyncClient;
import com.aliyun.sdk.service.starops20260428.models.CreateChatRequest;
import com.aliyun.sdk.service.starops20260428.models.CreateChatResponseBody;
import com.aliyun.sdk.service.starops20260428.models.CreateThreadRequest;
import com.aliyun.sdk.service.starops20260428.models.CreateThreadResponse;
import com.aliyun.sdk.service.starops20260428.models.DeleteThreadRequest;
import com.aliyun.sdk.service.starops20260428.models.GetThreadDataRequest;
import com.aliyun.sdk.service.starops20260428.models.GetThreadDataResponse;
import com.aliyun.sdk.service.starops20260428.models.GetThreadDataResponseBody;
import com.aliyun.sdk.service.starops20260428.models.GetThreadRequest;
import com.aliyun.sdk.service.starops20260428.models.GetThreadResponse;
import com.aliyun.sdk.service.starops20260428.models.GetThreadResponseBody;
import com.aliyun.sdk.service.starops20260428.models.ListThreadsRequest;
import com.aliyun.sdk.service.starops20260428.models.ListThreadsResponse;
import com.aliyun.sdk.service.starops20260428.models.ListThreadsResponseBody;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import darabonba.core.ResponseIterable;
import darabonba.core.client.ClientOverrideConfiguration;

/**
 * Agent 客户端（异步 SDK 版本，支持 SSE 流式输出）
 * Agent client using async SDK with SSE streaming support
 */
public class AgentClient implements AutoCloseable {
    private final AsyncClient client;
    private final Config config;
    private final ObjectMapper objectMapper;
    private final ExecutorService executor;

    public AgentClient(Config config) throws SDKException {
        this.config = config;
        this.objectMapper = new ObjectMapper();
        this.executor = Executors.newCachedThreadPool();

        try {
            StaticCredentialProvider provider = StaticCredentialProvider.create(
                    Credential.builder()
                            .accessKeyId(config.getAccessKeyId())
                            .accessKeySecret(config.getAccessKeySecret())
                            .build());

            this.client = AsyncClient.builder()
                    .credentialsProvider(provider)
                    .overrideConfiguration(
                            ClientOverrideConfiguration.create()
                                    .setEndpointOverride(config.getEndpoint()))
                    .serviceConfiguration(
                            Configuration.create()
                                    .setSignatureVersion(SignatureVersion.V3))
                    .build();
        } catch (Exception e) {
            throw SDKException.clientCreate(e);
        }
    }

    public Config getConfig() {
        return config;
    }

    /**
     * 创建会话
     * Create a new thread
     */
    public String createThread() throws SDKException {
        return createThread(null);
    }

    /**
     * 创建会话（支持自定义属性）
     * Create a new thread with optional attributes
     */
    public String createThread(Map<String, String> attributes) throws SDKException {
        try {
            CreateThreadRequest.Builder builder = CreateThreadRequest.builder()
                    .name(config.getEmployeeName())
                    .title("Chat-" + Instant.now().getEpochSecond())
                    .variables(CreateThreadRequest.Variables.builder()
                            .workspace(config.getWorkspace())
                            .build());

            if (attributes != null && !attributes.isEmpty()) {
                builder.attributes(attributes);
            }

            CreateThreadResponse response = client.createThread(builder.build()).get();

            if (response.getBody() == null || response.getBody().getThreadId() == null) {
                throw new SDKException(ErrorCode.THREAD_CREATE, "无效响应: 缺少ThreadID");
            }

            return response.getBody().getThreadId();
        } catch (SDKException e) {
            throw e;
        } catch (Exception e) {
            throw SDKException.threadCreate(e);
        }
    }

    /**
     * 开始 SSE 对话（基础版本）
     * Start SSE chat (basic version)
     */
    public BlockingQueue<ChatEvent> chat(String threadId, String message) {
        Map<String, Object> variables = new HashMap<>();
        variables.put("workspace", config.getWorkspace());
        variables.put("region", config.getRegion());
        variables.put("language", "zh");
        variables.put("timeZone", "Asia/Shanghai");
        variables.put("timeStamp", String.valueOf(Instant.now().getEpochSecond()));
        return chatWithVariables(threadId, message, variables);
    }

    /**
     * 开始对话（支持自定义 variables，使用异步 SDK 的 SSE 流式迭代）
     * Start chat with custom variables using async SDK's SSE streaming
     */
    public BlockingQueue<ChatEvent> chatWithVariables(String threadId, String message, Map<String, Object> variables) {
        BlockingQueue<ChatEvent> events = new LinkedBlockingQueue<>();

        final Map<String, Object> finalVariables = variables != null ? new HashMap<>(variables) : new HashMap<>();
        finalVariables.putIfAbsent("workspace", config.getWorkspace());
        finalVariables.putIfAbsent("region", config.getRegion());
        finalVariables.putIfAbsent("language", "zh");
        finalVariables.putIfAbsent("timeZone", "Asia/Shanghai");
        finalVariables.putIfAbsent("timeStamp", String.valueOf(Instant.now().getEpochSecond()));

        executor.submit(() -> {
            try {
                // 构建消息内容
                CreateChatRequest.Contents content = CreateChatRequest.Contents.builder()
                        .type("text")
                        .value(message)
                        .build();

                CreateChatRequest.Messages msg = CreateChatRequest.Messages.builder()
                        .role("user")
                        .contents(Collections.singletonList(content))
                        .build();

                CreateChatRequest request = CreateChatRequest.builder()
                        .regionId(config.getRegion())
                        .action("create")
                        .threadId(threadId)
                        .digitalEmployeeName(config.getEmployeeName())
                        .messages(Collections.singletonList(msg))
                        .variables(finalVariables)
                        .build();

                // 使用 SSE 流式迭代
                ResponseIterable<CreateChatResponseBody> iterable =
                        client.createChatWithResponseIterable(request);

                Iterator<CreateChatResponseBody> iterator = iterable.iterator();
                while (iterator.hasNext()) {
                    CreateChatResponseBody body = iterator.next();
                    if (body != null) {
                        // 将 ResponseBody 转为 JsonNode 以保持与现有 ChatEvent 兼容
                        Map<String, Object> map = bodyToMap(body);
                        String jsonStr = objectMapper.writeValueAsString(map);
                        JsonNode jsonNode = objectMapper.readTree(jsonStr);
                        ChatEvent event = ChatEvent.fromResponse(jsonNode, jsonStr, 200);
                        events.put(event);

                        if (event.isDone()) {
                            return;
                        }
                    }
                }

                events.put(ChatEvent.done());
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                try {
                    events.put(ChatEvent.error(SDKException.cancelled()));
                } catch (InterruptedException ignored) {}
            } catch (Exception e) {
                try {
                    events.put(ChatEvent.error(SDKException.chatFailed(e)));
                } catch (InterruptedException ignored) {}
            }
        });

        return events;
    }

    /**
     * 将 CreateChatResponseBody 转为 Map
     * Convert CreateChatResponseBody to Map for JSON serialization
     */
    private Map<String, Object> bodyToMap(CreateChatResponseBody body) {
        Map<String, Object> map = new LinkedHashMap<>();
        if (body.getRequestId() != null) {
            map.put("requestId", body.getRequestId());
        }
        if (body.getTraceId() != null) {
            map.put("traceId", body.getTraceId());
        }
        if (body.getMessages() != null) {
            List<Map<String, Object>> msgList = new ArrayList<>();
            for (CreateChatResponseBody.Messages msg : body.getMessages()) {
                Map<String, Object> msgMap = new LinkedHashMap<>();
                if (msg.getRole() != null) msgMap.put("role", msg.getRole());
                if (msg.getType() != null) msgMap.put("type", msg.getType());
                if (msg.getCallId() != null) msgMap.put("callId", msg.getCallId());
                if (msg.getParentCallId() != null) msgMap.put("parentCallId", msg.getParentCallId());
                if (msg.getSeq() != null) msgMap.put("seq", msg.getSeq());
                if (msg.getTimestamp() != null) msgMap.put("timestamp", msg.getTimestamp());
                if (msg.getVersion() != null) msgMap.put("version", msg.getVersion());
                if (msg.getDetail() != null) msgMap.put("detail", msg.getDetail());
                if (msg.getContents() != null) msgMap.put("contents", msg.getContents());
                if (msg.getArtifacts() != null) msgMap.put("artifacts", msg.getArtifacts());
                if (msg.getTools() != null) msgMap.put("tools", msg.getTools());
                if (msg.getAgents() != null) msgMap.put("agents", msg.getAgents());
                if (msg.getEvents() != null) msgMap.put("events", msg.getEvents());
                msgList.add(msgMap);
            }
            map.put("messages", msgList);
        }
        return map;
    }

    /**
     * 带超时的对话
     * Chat with timeout
     */
    public BlockingQueue<ChatEvent> chatWithTimeout(String threadId, String message, java.time.Duration timeout) {
        BlockingQueue<ChatEvent> events = chat(threadId, message);
        BlockingQueue<ChatEvent> timedEvents = new LinkedBlockingQueue<>();

        executor.submit(() -> {
            long deadline = System.currentTimeMillis() + timeout.toMillis();
            try {
                while (true) {
                    long remaining = deadline - System.currentTimeMillis();
                    if (remaining <= 0) {
                        timedEvents.put(ChatEvent.error(SDKException.timeout(timeout.toString())));
                        return;
                    }

                    ChatEvent event = events.poll(remaining, TimeUnit.MILLISECONDS);
                    if (event == null) {
                        timedEvents.put(ChatEvent.error(SDKException.timeout(timeout.toString())));
                        return;
                    }

                    timedEvents.put(event);
                    if (event.isDone() || event.hasError()) {
                        return;
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                try {
                    timedEvents.put(ChatEvent.error(SDKException.cancelled()));
                } catch (InterruptedException ignored) {}
            }
        });

        return timedEvents;
    }

    /**
     * 列出会话
     * List threads
     */
    public ListThreadsResult listThreads(int pageSize) throws SDKException {
        try {
            if (pageSize <= 0) pageSize = 20;
            if (pageSize > 100) pageSize = 100;

            ListThreadsRequest request = ListThreadsRequest.builder()
                    .name(config.getEmployeeName())
                    .maxResults((long) pageSize)
                    .build();

            ListThreadsResponse response = client.listThreads(request).get();

            if (response.getBody() == null) {
                throw new SDKException(ErrorCode.PARSE_ERROR, "无效响应: 响应体为空")
                        .withSuggestion("请稍后重试");
            }

            List<ThreadInfo> threads = new ArrayList<>();
            ListThreadsResponseBody body = response.getBody();
            if (body.getThreads() != null) {
                for (ListThreadsResponseBody.Threads t : body.getThreads()) {
                    threads.add(new ThreadInfo(
                            t.getThreadId(),
                            t.getTitle(),
                            t.getStatus(),
                            t.getCreateTime(),
                            t.getUpdateTime()
                    ));
                }
            }

            long total = body.getTotal() != null ? body.getTotal() : 0;
            return new ListThreadsResult(threads, total);
        } catch (SDKException e) {
            throw e;
        } catch (Exception e) {
            throw new SDKException(ErrorCode.API_ERROR, "获取会话列表失败", e)
                    .withSuggestion("请检查网络连接和 API 权限");
        }
    }

    /**
     * 获取会话详情
     * Get thread details
     */
    public ThreadInfo getThread(String threadId) throws SDKException {
        validateThreadId(threadId);

        try {
            GetThreadRequest request = GetThreadRequest.builder()
                    .name(config.getEmployeeName())
                    .threadId(threadId)
                    .build();

            GetThreadResponse response = client.getThread(request).get();

            if (response.getBody() == null) {
                throw new SDKException(ErrorCode.PARSE_ERROR, "无效响应: 响应体为空")
                        .withContext("threadId", threadId)
                        .withSuggestion("请稍后重试");
            }

            GetThreadResponseBody body = response.getBody();
            return new ThreadInfo(
                    body.getThreadId(),
                    body.getTitle(),
                    body.getStatus(),
                    body.getCreateTime(),
                    body.getUpdateTime()
            );
        } catch (SDKException e) {
            throw e;
        } catch (Exception e) {
            if (isThreadNotFoundError(e)) {
                throw SDKException.threadNotFound(threadId);
            }
            throw new SDKException(ErrorCode.API_ERROR, "获取会话详情失败: " + threadId, e)
                    .withContext("threadId", threadId)
                    .withSuggestion("请检查会话 ID 是否正确");
        }
    }

    /**
     * 删除会话
     * Delete thread
     */
    public void deleteThread(String threadId) throws SDKException {
        validateThreadId(threadId);

        try {
            DeleteThreadRequest request = DeleteThreadRequest.builder()
                    .name(config.getEmployeeName())
                    .threadId(threadId)
                    .build();

            client.deleteThread(request).get();
        } catch (Exception e) {
            if (isThreadNotFoundError(e)) {
                throw SDKException.threadNotFound(threadId);
            }
            throw new SDKException(ErrorCode.API_ERROR, "删除会话失败: " + threadId, e)
                    .withContext("threadId", threadId)
                    .withSuggestion("请检查会话 ID 是否正确");
        }
    }

    /**
     * 获取会话消息
     * Get thread messages
     */
    public List<ThreadMessage> getThreadData(String threadId, int limit) throws SDKException {
        validateThreadId(threadId);

        try {
            if (limit <= 0) limit = 50;
            if (limit > 100) limit = 100;

            GetThreadDataRequest request = GetThreadDataRequest.builder()
                    .name(config.getEmployeeName())
                    .threadId(threadId)
                    .maxResults((long) limit)
                    .build();

            GetThreadDataResponse response = client.getThreadData(request).get();

            if (response.getBody() == null) {
                throw new SDKException(ErrorCode.PARSE_ERROR, "无效响应: 响应体为空")
                        .withContext("threadId", threadId)
                        .withSuggestion("请稍后重试");
            }

            GetThreadDataResponseBody body = response.getBody();

            // 策略：优先使用 system Result，而不是 assistant 流式消息
            Map<String, ThreadMessage> messageMap = new LinkedHashMap<>();

            // 检查是否存在 system Result
            boolean hasSystemResult = false;
            if (body.getData() != null) {
                outer:
                for (GetThreadDataResponseBody.Data data : body.getData()) {
                    if (data.getMessages() != null) {
                        for (GetThreadDataResponseBody.Messages msg : data.getMessages()) {
                            if ("system".equals(msg.getRole()) && msg.getArtifacts() != null) {
                                for (Map<String, ?> artifact : msg.getArtifacts()) {
                                    if (artifact != null && "Result".equals(artifact.get("name"))) {
                                        hasSystemResult = true;
                                        break outer;
                                    }
                                }
                            }
                        }
                    }
                }
            }

            // 处理消息
            if (body.getData() != null) {
                for (GetThreadDataResponseBody.Data data : body.getData()) {
                    if (data.getMessages() != null) {
                        for (GetThreadDataResponseBody.Messages msg : data.getMessages()) {
                            String role = msg.getRole() != null ? msg.getRole() : "";
                            String timestamp = msg.getTimestamp() != null ? msg.getTimestamp() : "";

                            // 如果存在 system Result，跳过 assistant 流式消息
                            if ("assistant".equals(role) && hasSystemResult) {
                                continue;
                            }

                            String key;
                            if ("user".equals(role)) {
                                key = "user-" + timestamp;
                            } else if ("system".equals(role)) {
                                key = "system-" + timestamp;
                            } else {
                                String callId = msg.getCallId() != null ? msg.getCallId() : "";
                                key = "assistant-" + callId;
                            }

                            String content = extractMessageContent(msg);
                            if (content == null || content.isEmpty()) {
                                continue;
                            }

                            if (messageMap.containsKey(key)) {
                                ThreadMessage existing = messageMap.get(key);
                                messageMap.put(key, new ThreadMessage(
                                        existing.getRole(),
                                        existing.getContent() + content,
                                        existing.getTimestamp()
                                ));
                            } else {
                                String displayRole = "system".equals(role) ? "assistant" : role;
                                messageMap.put(key, new ThreadMessage(displayRole, content, timestamp));
                            }
                        }
                    }
                }
            }

            return new ArrayList<>(messageMap.values());
        } catch (SDKException e) {
            throw e;
        } catch (Exception e) {
            if (isThreadNotFoundError(e)) {
                throw SDKException.threadNotFound(threadId);
            }
            throw new SDKException(ErrorCode.API_ERROR, "获取会话消息失败: " + threadId, e)
                    .withContext("threadId", threadId)
                    .withSuggestion("请检查会话 ID 是否正确");
        }
    }

    private void validateThreadId(String threadId) throws SDKException {
        if (threadId == null || threadId.isEmpty()) {
            throw new SDKException(ErrorCode.CONFIG_INVALID, "会话 ID 不能为空")
                    .withContext("threadId", threadId)
                    .withSuggestion("请提供有效的会话 ID");
        }
        if (threadId.contains(" ") || threadId.contains("\t") || threadId.contains("\n")) {
            throw new SDKException(ErrorCode.CONFIG_INVALID, "会话 ID 包含非法字符: " + threadId)
                    .withContext("threadId", threadId)
                    .withSuggestion("会话 ID 不能包含空白字符");
        }
    }

    private boolean isThreadNotFoundError(Exception e) {
        if (e == null) return false;
        String errStr = e.getMessage();
        if (errStr == null) return false;
        return errStr.contains("NotFound") || errStr.contains("not found") ||
               errStr.contains("NOT_FOUND") || errStr.contains("ThreadNotFound") ||
               errStr.contains("InvalidThreadId") || errStr.contains("does not exist");
    }

    @SuppressWarnings("unchecked")
    private String extractMessageContent(GetThreadDataResponseBody.Messages msg) {
        if (msg == null) return "";

        // 1. 尝试从 contents 提取（流式文本块）
        if (msg.getContents() != null) {
            StringBuilder result = new StringBuilder();
            for (Map<String, ?> content : msg.getContents()) {
                if (content != null) {
                    Object type = content.get("type");
                    Object value = content.get("value");
                    if ("text".equals(type) && value != null) {
                        result.append(value.toString());
                    }
                }
            }
            if (result.length() > 0) {
                return result.toString();
            }
        }

        // 2. 尝试从 artifacts 提取（最终结果）
        if (msg.getArtifacts() != null) {
            for (Map<String, ?> artifact : msg.getArtifacts()) {
                if (artifact != null && "Result".equals(artifact.get("name"))) {
                    Object partsObj = artifact.get("parts");
                    if (partsObj instanceof List) {
                        List<?> parts = (List<?>) partsObj;
                        StringBuilder textParts = new StringBuilder();
                        for (Object partObj : parts) {
                            if (partObj instanceof Map) {
                                Map<?, ?> part = (Map<?, ?>) partObj;
                                if ("text".equals(part.get("kind")) && part.get("text") != null) {
                                    textParts.append(part.get("text").toString());
                                }
                            }
                        }
                        if (textParts.length() > 0) {
                            return textParts.toString();
                        }
                    }
                }
            }
        }

        return "";
    }

    public void shutdown() {
        executor.shutdown();
        try {
            client.close();
        } catch (Exception ignored) {}
    }

    @Override
    public void close() {
        shutdown();
    }

    /**
     * 列出会话结果
     * List threads result
     */
    public static class ListThreadsResult {
        private final List<ThreadInfo> threads;
        private final long total;

        public ListThreadsResult(List<ThreadInfo> threads, long total) {
            this.threads = threads;
            this.total = total;
        }

        public List<ThreadInfo> getThreads() {
            return threads;
        }

        public long getTotal() {
            return total;
        }
    }
}

File: samples/typescript/src/client/agent-client.ts
/**
 * Agent client for CMS SDK
 * CMS SDK Agent 客户端
 */

import CMS20240330Module, * as $CMS20240330 from '@alicloud/cms20240330';
import * as $OpenApi from '@alicloud/openapi-client';
import * as $dara from '@darabonba/typescript';
import { Config } from './config.js';
import { SDKException, ErrorCode } from './errors.js';

// Handle ESM/CJS interop
const CMS20240330Client = (CMS20240330Module as any).default || CMS20240330Module;

/** 聊天事件 / Chat event */
export interface ChatEvent {
  body?: Record<string, unknown>;
  rawJson: string;
  statusCode: number;
  isDone: boolean;
  error?: Error;
}

/** 会话信息 / Thread information */
export interface ThreadInfo {
  threadId: string;
  title: string;
  status: string;
  createTime: string;
  updateTime: string;
}

/** 会话消息 / Thread message */
export interface ThreadMessage {
  role: string;
  content: string;
  timestamp: string;
}

function isDoneMessage(body?: Record<string, unknown>): boolean {
  if (!body || !body.messages) return false;
  const messages = body.messages as Array<Record<string, unknown>>;
  return messages.some((msg) => msg.type === 'done');
}

/** Agent 客户端 / Agent client */
export class AgentClient {
  private client: any;
  private config: Config;

  constructor(config: Config) {
    this.config = config;

    try {
      const openApiConfig = new $OpenApi.Config({
        accessKeyId: config.accessKeyId,
        accessKeySecret: config.accessKeySecret,
        endpoint: config.endpoint,
        signatureVersion: 'v3',
      });
      this.client = new CMS20240330Client(openApiConfig);
    } catch (e) {
      throw SDKException.clientCreate(e as Error);
    }
  }

  getConfig(): Config {
    return this.config;
  }

  /** 创建会话 / Create thread */
  async createThread(): Promise<string> {
    try {
      const variables = new $CMS20240330.CreateThreadRequestVariables({
        workspace: this.config.workspace,
      });
      const request = new $CMS20240330.CreateThreadRequest({
        title: `Chat-${Math.floor(Date.now() / 1000)}`,
        variables,
      });

      const response = await this.client.createThread(
        this.config.employeeName,
        request
      );

      if (!response.body?.threadId) {
        throw new SDKException(ErrorCode.THREAD_CREATE, '无效响应: 缺少ThreadID');
      }

      return response.body.threadId;
    } catch (e) {
      if (e instanceof SDKException) throw e;
      throw SDKException.threadCreate(e as Error);
    }
  }

  /** 开始 SSE 对话 / Start SSE chat */
  async *chat(threadId: string, message: string): AsyncIterable<ChatEvent> {
    const variables: Record<string, unknown> = {
      workspace: this.config.workspace,
      region: this.config.region,
      language: 'zh',
      timeZone: 'Asia/Shanghai',
      timeStamp: String(Math.floor(Date.now() / 1000)),
    };
    yield* this.chatWithVariables(threadId, message, variables);
  }

  /** 开始 SSE 对话（支持自定义 variables）/ Start SSE chat with custom variables */
  async *chatWithVariables(
    threadId: string,
    message: string,
    variables?: Record<string, unknown>
  ): AsyncIterable<ChatEvent> {
    try {
      const content = new $CMS20240330.CreateChatRequestMessagesContents({
        type: 'text',
        value: message,
      });
      const msg = new $CMS20240330.CreateChatRequestMessages({
        role: 'user',
        contents: [content],
      });

      // Ensure required fields
      const vars = variables || {};
      vars.workspace = vars.workspace || this.config.workspace;
      vars.region = vars.region || this.config.region;
      vars.language = vars.language || 'zh';
      vars.timeZone = vars.timeZone || 'Asia/Shanghai';
      vars.timeStamp = vars.timeStamp || String(Math.floor(Date.now() / 1000));

      const request = new $CMS20240330.CreateChatRequest({
        action: 'create',
        threadId,
        digitalEmployeeName: this.config.employeeName,
        messages: [msg],
        variables: vars,
      });

      const runtime = new $dara.RuntimeOptions({
        readTimeout: 120000,  // 2 minutes
        connectTimeout: 30000, // 30 seconds
      });
      const responseIterator = await this.client.createChatWithSSE(request, {}, runtime);

      for await (const response of responseIterator) {
        if (response.body) {
          const bodyObj = response.body as unknown as Record<string, unknown>;
          const rawJson = JSON.stringify(bodyObj);
          const event: ChatEvent = {
            body: bodyObj,
            rawJson,
            statusCode: 200,
            isDone: isDoneMessage(bodyObj),
          };
          yield event;
          if (event.isDone) return;
        }
      }

      yield { rawJson: '', statusCode: 200, isDone: true };
    } catch (e) {
      const err = e as Error;
      const sdkErr = SDKException.chatFailed(err);
      // Preserve original error details
      if (err.message) {
        sdkErr.withContext('originalMessage', err.message);
      }
      if ((err as any).code) {
        sdkErr.withContext('errorCode', (err as any).code);
      }
      if ((err as any).data) {
        sdkErr.withContext('errorData', JSON.stringify((err as any).data));
      }
      yield { rawJson: '', statusCode: 0, isDone: false, error: sdkErr };
    }
  }

  /** 带超时的对话 / Chat with timeout */
  async *chatWithTimeout(
    threadId: string,
    message: string,
    timeout: number
  ): AsyncIterable<ChatEvent> {
    const deadline = Date.now() + timeout;

    for await (const event of this.chat(threadId, message)) {
      if (Date.now() > deadline) {
        yield { rawJson: '', statusCode: 0, isDone: false, error: SDKException.timeout(`${timeout}ms`) };
        return;
      }
      yield event;
      if (event.isDone || event.error) return;
    }
  }

  /** 列出会话 / List threads */
  async listThreads(pageSize = 20): Promise<{ threads: ThreadInfo[]; total: number }> {
    try {
      if (pageSize <= 0) pageSize = 20;
      if (pageSize > 100) pageSize = 100;

      const request = new $CMS20240330.ListThreadsRequest({
        maxResults: pageSize,
      });

      const response = await this.client.listThreads(
        this.config.employeeName,
        request
      );

      if (!response.body) {
        throw new SDKException(ErrorCode.PARSE_ERROR, '无效响应: 响应体为空')
          .withSuggestion('请稍后重试');
      }

      const threads: ThreadInfo[] = (response.body.threads || []).map((t: any) => ({
        threadId: t.threadId || '',
        title: t.title || '',
        status: t.status || '',
        createTime: t.createTime || '',
        updateTime: t.updateTime || '',
      }));

      return { threads, total: response.body.total || 0 };
    } catch (e) {
      if (e instanceof SDKException) throw e;
      throw new SDKException(ErrorCode.API_ERROR, '获取会话列表失败', e as Error)
        .withSuggestion('请检查网络连接和 API 权限');
    }
  }

  /** 获取会话详情 / Get thread details */
  async getThread(threadId: string): Promise<ThreadInfo> {
    this.validateThreadId(threadId);

    try {
      const response = await this.client.getThread(
        this.config.employeeName,
        threadId
      );

      if (!response.body) {
        throw new SDKException(ErrorCode.PARSE_ERROR, '无效响应: 响应体为空')
          .withContext('threadId', threadId)
          .withSuggestion('请稍后重试');
      }

      return {
        threadId: response.body.threadId || '',
        title: response.body.title || '',
        status: response.body.status || '',
        createTime: response.body.createTime || '',
        updateTime: response.body.updateTime || '',
      };
    } catch (e) {
      if (e instanceof SDKException) throw e;
      if (this.isThreadNotFoundError(e as Error)) {
        throw SDKException.threadNotFound(threadId);
      }
      throw new SDKException(ErrorCode.API_ERROR, `获取会话详情失败: ${threadId}`, e as Error)
        .withContext('threadId', threadId)
        .withSuggestion('请检查会话 ID 是否正确');
    }
  }

  /** 删除会话 / Delete thread */
  async deleteThread(threadId: string): Promise<void> {
    this.validateThreadId(threadId);

    try {
      await this.client.deleteThread(this.config.employeeName, threadId);
    } catch (e) {
      if (this.isThreadNotFoundError(e as Error)) {
        throw SDKException.threadNotFound(threadId);
      }
      throw new SDKException(ErrorCode.API_ERROR, `删除会话失败: ${threadId}`, e as Error)
        .withContext('threadId', threadId)
        .withSuggestion('请检查会话 ID 是否正确');
    }
  }

  /** 获取会话消息 / Get thread messages */
  async getThreadData(threadId: string, limit = 50): Promise<ThreadMessage[]> {
    this.validateThreadId(threadId);

    try {
      if (limit <= 0) limit = 50;
      if (limit > 100) limit = 100;

      const request = new $CMS20240330.GetThreadDataRequest({
        maxResults: limit,
      });

      const response = await this.client.getThreadData(
        this.config.employeeName,
        threadId,
        request
      );

      if (!response.body) {
        throw new SDKException(ErrorCode.PARSE_ERROR, '无效响应: 响应体为空')
          .withContext('threadId', threadId)
          .withSuggestion('请稍后重试');
      }

      // Strategy: prefer system Result over assistant streaming messages
      // 策略：优先使用 system Result，而不是 assistant 流式消息
      // - user messages: use timestamp as key
      // - system messages with Result: use timestamp as key (display as assistant)
      // - assistant messages: only include if no system Result exists
      
      const messageMap = new Map<string, { role: string; content: string; timestamp: string }>();
      const messageOrder: string[] = [];
      
      // Check if any system Result exists
      // 检查是否存在 system Result
      let hasSystemResult = false;
      for (const data of response.body.data || []) {
        for (const msg of data.messages || []) {
          if (msg.role === 'system') {
            const artifacts = (msg as any).artifacts as Array<Record<string, unknown>> | undefined;
            if (artifacts?.some(a => a.name === 'Result')) {
              hasSystemResult = true;
              break;
            }
          }
        }
        if (hasSystemResult) break;
      }

      // Process messages
      // 处理消息
      for (const data of response.body.data || []) {
        for (const msg of data.messages || []) {
          const role = msg.role || '';
          const timestamp = msg.timestamp || '';
          
          // Skip assistant streaming messages if system Result exists
          // 如果存在 system Result，跳过 assistant 流式消息
          if (role === 'assistant' && hasSystemResult) {
            continue;
          }
          
          // Use different key strategy based on role
          // 根据角色使用不同的键策略
          let key: string;
          if (role === 'user') {
            key = `user-${timestamp}`;
          } else if (role === 'system') {
            key = `system-${timestamp}`;
          } else {
            const callId = (msg as any).callId || '';
            key = `assistant-${callId}`;
          }
          
          const content = this.extractMessageContent(msg);
          if (!content) continue;

          if (messageMap.has(key)) {
            // Append content to existing message
            const existing = messageMap.get(key)!;
            existing.content += content;
          } else {
            // For system messages, display as assistant role
            // 对于 system 消息，显示为 assistant 角色
            const displayRole = role === 'system' ? 'assistant' : role;
            messageMap.set(key, { role: displayRole, content, timestamp });
            messageOrder.push(key);
          }
        }
      }

      // Return messages in order
      return messageOrder.map(key => messageMap.get(key)!);
    } catch (e) {
      if (e instanceof SDKException) throw e;
      if (this.isThreadNotFoundError(e as Error)) {
        throw SDKException.threadNotFound(threadId);
      }
      throw new SDKException(ErrorCode.API_ERROR, `获取会话消息失败: ${threadId}`, e as Error)
        .withContext('threadId', threadId)
        .withSuggestion('请检查会话 ID 是否正确');
    }
  }

  private validateThreadId(threadId: string): void {
    if (!threadId) {
      throw new SDKException(ErrorCode.CONFIG_INVALID, '会话 ID 不能为空')
        .withContext('threadId', threadId)
        .withSuggestion('请提供有效的会话 ID');
    }
    if (/[\s\t\n\r]/.test(threadId)) {
      throw new SDKException(ErrorCode.CONFIG_INVALID, `会话 ID 包含非法字符: ${threadId}`)
        .withContext('threadId', threadId)
        .withSuggestion('会话 ID 不能包含空白字符');
    }
  }

  private isThreadNotFoundError(e: Error): boolean {
    if (!e) return false;
    const errStr = e.message || '';
    const patterns = ['NotFound', 'not found', 'NOT_FOUND', 'ThreadNotFound', 'InvalidThreadId', 'does not exist'];
    return patterns.some((p) => errStr.includes(p));
  }

  private extractMessageContent(msg: Record<string, unknown>): string {
    // 1. Try to extract from contents (streaming text chunks)
    // 尝试从 contents 提取（流式文本块）
    const contents = msg.contents as Array<Record<string, unknown>> | undefined;
    if (contents) {
      const textContent = contents
        .filter((c) => c.type === 'text' && c.value)
        .map((c) => c.value as string)
        .join('');
      if (textContent) return textContent;
    }

    // 2. Try to extract from artifacts (final result)
    // 尝试从 artifacts 提取（最终结果）
    const artifacts = msg.artifacts as Array<Record<string, unknown>> | undefined;
    if (artifacts) {
      for (const artifact of artifacts) {
        // Look for Result artifact with text parts
        // 查找包含文本的 Result artifact
        if (artifact.name === 'Result') {
          const parts = artifact.parts as Array<Record<string, unknown>> | undefined;
          if (parts) {
            const textParts = parts
              .filter((p) => p.kind === 'text' && p.text)
              .map((p) => p.text as string);
            if (textParts.length > 0) return textParts.join('');
          }
        }
      }
    }

    return '';
  }
}
